### PR TITLE
Implemented prototype replace from file for replacement editor

### LINKED ISSUE #1458: Add Replace from file to Replacement Editor
We have a replacement editor (press `o` then `R` in mitmproxy) that allows us to add replacement patterns for flows. 

There's also a `--replace-from-file` cmdline option that gets the replacement text from a local file.

But this functionality can't be used from the replacement editor itself. 

---

A user tried to enter the path to file in the Replacement text field: https://discourse.mitmproxy.org/t/emulate-charles-proxys-map-local/116

### PR DESCRIPTION
**Pull request by arjun23496**
fixes #1458

### CODE REVIEW

File: mitmproxy/builtins/replace.py [hunk 1]
```diff
@@ -29,16 +31,30 @@ def configure(self, options, updated):
                 raise exceptions.OptionsError(
                     "Invalid regular expression: %s - %s" % (rex, str(e))
                 )
-            lst.append((rex, s, cpatt))
+            if s[0:7] == "file://":
```
Comment (Kriechi) [inline mitmproxy/builtins/replace.py]: Please use `s.startswith('file://')` instead.

File: mitmproxy/builtins/replace.py [hunk 2]
```diff
@@ -29,16 +31,30 @@ def configure(self, options, updated):
...
+                file_repl = True
+                if not os.path.isfile(s[7:]):
```
Comment (Kriechi) [inline mitmproxy/builtins/replace.py]: Should we use https://docs.python.org/3/library/urllib.parse.html#module-urllib.parse to properly parse file urls?
Might be safer...

File: mitmproxy/builtins/replace.py [hunk 3]
```diff
@@ -29,16 +31,30 @@ def configure(self, options, updated):
...
+            lst.append((rex, s, cpatt, file_repl))
         self.lst = lst
 
     def execute(self, f):
-        for rex, s, cpatt in self.lst:
+        for rex, s, cpatt, file_repl in self.lst:
             if cpatt(f):
-                if f.response:
-                    f.response.replace(rex, s)
+                if not file_repl:
```
Comment (Kriechi) [inline mitmproxy/builtins/replace.py]: Please flip the two if-else blocks, so we do not need the negation in the condition.

File: mitmproxy/builtins/replace.py [hunk 4]
```diff
@@ -29,16 +31,30 @@ def configure(self, options, updated):
                 raise exceptions.OptionsError(
                     "Invalid regular expression: %s - %s" % (rex, str(e))
                 )
-            lst.append((rex, s, cpatt))
+            if s[0:7] == "file://":
+                file_repl = True
+                if not os.path.isfile(s[7:]):
+                    raise exceptions.OptionsError(
+                        "File not found : %s" % s[7:]
```
Comment (Kriechi) [inline mitmproxy/builtins/replace.py]: Please remove the space before the colon, and use the full `s` variable (not `[7:]`)

File: netlib/http/message.py [hunk 5]
```diff
@@ -260,6 +260,26 @@ def encode(self, e):
...
+            if isinstance(repl_chunk, six.text_type):
+                repl_chunk = strutils.escaped_str_to_bytes(repl_chunk)
+            if self.content:
+                self.content, replacements = re.subn(
+                    pattern, repl_chunk, self.content, flags=flags, count=count
+                )
+            replacements += self.headers.replace(pattern, repl_chunk, flags=flags, count=count)
+            return replacements
+
+        return replacements
```
Comment (Kriechi) [inline netlib/http/message.py]: Doesn't the return 2 lines above already exit this function? Or is there some error handling missing?

**Comment by Kriechi** [comment]
Sorry this took so long to review - seems to have slipped through the crack.
Please address the inline comments above, and resolve the conflicts.

**Comment by cortesi** [comment]
Superseded by #2120. Arjun, please don't be discouraged by the fact that you got pipped to the post this time. We hope to see more PRs from you in future.

**Pull request closed without merging**